### PR TITLE
UC2: What needs to happen for me to win? (Closes #25-#29)

### LINKED ISSUE #25: Use Case 2: "What needs to happen for me to win?"
## Problem

After checking the standings, every player immediately asks: *"Can I still win, and what has to happen?"* Today that requires navigating two tabs, running manual what-if simulations, and mentally cross-referencing critical games against your own picks. The app should give a direct, plain-English answer: here are the specific outcomes that give you the best shot, and here's how likely that path is.

## Requirements

Full product requirements with user stories and acceptance criteria are written up in the session. Key asks:

- **Player-centric critical games** — re-rank by *my* personal swing, not the global max swing
- **Must-win games** — flag games where losing drops my win probability to 0%
- **Best-case path** — plain-English "root for X to beat Y" list for the outcome that maximizes my win %
- **Path probability** — how likely is that best-case path, based on current odds?
- **Clinch detection** — does any combination of outcomes *guarantee* me 1st place?
- **Player selector** — switch perspective between players; defaults to `my_player_name` from config
- **Eliminated state** — suppress path analysis and show clear message if player can't win

## Work Items

| WI | Issue | Effort | Priority | Depends On | Track |
|----|-------|--------|----------|------------|-------|
| UC2-1 | #26 Core — `player_critical_games()` | 1-2 hrs | Must-have | None | A |
| UC2-2 | #27 Core — `clinch_scenarios()` | 2-3 hrs | Must-have | None | A |
| UC2-3 | #28 Core — `best_path()` | 2-3 hrs | Should-have | #26 | B |
| UC2-4 | #29 Plugin UI — `win_probability.py` | 3-4 hrs | Must-have | #26, #27, #28 | C |

## Dependency Graph

```
#26 UC2-1 player_critical_games()   ← no dependencies, start here
#27 UC2-2 clinch_scenarios()        ← no dependencies, run parallel with #26
  └── #28 UC2-3 best_path()         ← needs #26 for game probabilities
        └── #29 UC2-4 Plugin UI     ← needs all three above
```

## Parallel Tracks

| Track | Work Items | Focus |
|-------|------------|-------|
| **A**: Core helpers | #26, #27 | Scenario math, deterministic clinch logic |
| **B**: Path logic | #28 | Odds integration, path probability |
| **C**: UI | #29 | Streamlit rendering, player selector |

## Coordination Note

#29 adds a player selector to `analyses/win_probability.py`. When issue #22 (WI-5 counterfactual UI) is built, its player selector in `analyses/scenarios_whatif.py` should follow the same pattern and default-to-`my_player_name` behavior.

## No conflicts with in-flight work

- Counterfactual issues (#16–22) touch `core/comparison.py` and `analyses/scenarios_whatif.py` only
- UC2 work touches `core/scenarios.py` (new helpers) and `analyses/win_probability.py` only
- Safe to implement in parallel

## Branch

`feature/uc2-path-to-win`

### LINKED ISSUE #26: UC2-1: Core — player_critical_games()
## Summary

Re-rank `ScenarioResults.critical_games` by a specific player's personal swing rather than the global `max_swing`. Also flag "must-win" games where the player's win probability drops to 0% if their pick loses.

## What to build

```python
# core/scenarios.py

def player_critical_games(
    sr: ScenarioResults,
    player_name: str,
    top_n: int = 3,
) -> list[dict]:
    """Re-rank critical games by a single player's personal win probability swing.

    Returns list of dicts (top_n), sorted by player's swing descending:
      - slot_id: str
      - team_a: str
      - team_b: str
      - win_if_a: float  (player's win % if team_a wins)
      - win_if_b: float  (player's win % if team_b wins)
      - swing: float     (absolute difference)
      - must_win_team: str | None  (team player needs; set if one side → 0.0%)
    """
```

## Implementation notes

- `sr.critical_games` already contains per-player tuples in `swings[player_name] = (win_if_a, win_if_b)`
- Extract each game's swing for the target player, re-sort descending
- `must_win_team`: if `win_if_a == 0.0`, the player must have team_b win (and vice versa)
- Return only games where the player has a non-zero swing (games that don't affect them at all can be omitted)

## Key files

- `core/scenarios.py` — add function here alongside existing scenario helpers
- `tests/test_scenarios.py` — add tests

## Acceptance criteria

- [ ] Returns games sorted by selected player's swing, not global `max_swing`
- [ ] `must_win_team` is set correctly when one outcome eliminates the player
- [ ] Games with zero swing for the player are excluded
- [ ] Works when player is already eliminated (returns empty list)
- [ ] `top_n` parameter respected

## Branch

`feature/uc2-path-to-win`

### LINKED ISSUE #27: UC2-2: Core — clinch_scenarios()
## Summary

Deterministic clinch detection: find the specific set of game outcomes that guarantees a player 1st place regardless of all other results. If no clinch exists, return the minimum number of remaining picks that must be correct for the player to have any win probability > 0%.

## What to build

```python
# core/scenarios.py

def clinch_scenarios(
    entries: list[PlayerEntry],
    player_name: str,
    tournament: TournamentStructure,
    results: Results,
) -> dict:
    """Detect clinch scenarios and elimination thresholds for a player.

    Returns dict with:
      - clinched: bool  (True if player has already clinched mathematically)
      - clinch_outcomes: list[dict] | None
            Each dict: {slot_id, required_winner} — outcomes that guarantee 1st
            None if no clinch scenario exists
      - can_win: bool  (False if win_prob == 0 regardless of outcomes)
      - min_picks_needed: int  (minimum additional correct picks for any win path)
    """
```

### Clinch check (already clinched)
- Player is clinched if `scored.total_points > max(other.max_possible for all others)`
- This is deterministic and cheap — check first

### Clinch scenario (can clinch with specific outcomes)
- For each remaining game, the player's pick either wins or loses
- Build a hypothetical Results where all of the player's remaining picks win
- Check if the player's resulting score exceeds all others' max possible in that scenario
- If yes: those picks constitute a clinch scenario — return the required outcomes
- If no clinch exists: return `clinch_outcomes=None`

### Minimum picks needed
- Iterate from 0 upward: how many correct picks does the player need before any win probability > 0% becomes possible?
- Use `max_possible` arithmetic: find the minimum correct-picks subset where player's ceiling > all others' floors
- Return that count as `min_picks_needed`

## Key files

- `core/scenarios.py` — add function
- `core/scoring.py` — `score_entry()`, `POINTS_PER_ROUND` for max possible arithmetic
- `tests/test_scenarios.py` — add tests

## Acceptance criteria

- [ ] Returns `clinched=True` when player's current score already exceeds all others' max possible
- [ ] Returns correct `clinch_outcomes` list (specific slot + required winner) when a clinch path exists
- [ ] Returns `clinch_outcomes=None` when no combination of outcomes can guarantee 1st
- [ ] Returns `can_win=False` when player is fully eliminated
- [ ] `min_picks_needed` is 0 when already clinched, correct count otherwise

## Branch

`feature/uc2-path-to-win`

### LINKED ISSUE #28: UC2-3: Core — best_path()
## Summary

Find the combination of remaining game outcomes that maximizes a specific player's win probability, then compute the real-world likelihood of that path using current betting odds.

## What to build

```python
# core/scenarios.py

def best_path(
    sr: ScenarioResults,
    player_name: str,
    tournament: TournamentStructure,
    results: Results,
    odds: dict | None = None,
) -> dict:
    """Find the outcome combination that maximizes a player's win probability.

    Returns dict with:
      - steps: list[dict]  — ordered list of {slot_id, round, root_for, opponent}
      - win_probability: float  — player's win % along this path
      - path_probability: float  — product of individual game odds (0.0–1.0)
      - odds_source: str  — "moneyline" | "seed_historical" | "coin_flip" etc.
    """
```

### Finding the best path
- From `sr.finish_distributions` and `sr.win_counts`, identify which scenarios the player wins
- Use `player_critical_games()` (UC2-1) to get the top games ranked by the player's swing
- The best path is the combination of outcomes (one per remaining game where player has a pick alive) that, taken together, yields the highest conditional win probability
- For small N (≤ 15 games, brute-force engine): enumerate wins directly from scenario data
- For large N (Monte Carlo): use top-swing games greedily — pick the outcome that most improves win % for each game in order

### Path probability
- For each step in the path, call `get_game_probability(team_a, team_b, tournament, odds, slot_id)`
- Path probability = product of `prob` for each required outcome
- Track the lowest-confidence source across all steps for `odds_source`

### Presentation order
- Sort steps by round ascending (earlier games first — they unlock later rounds)
- Only include games where the player's pick is still alive (no point listing eliminated teams)

## Key files

- `core/scenarios.py` — add function, reuse `get_game_probability()`
- `core/tournament.py` — `get_remaining_slots()` for pending games
- `tests/test_scenarios.py` — add tests

## Acceptance criteria

- [ ] Returns steps in round-ascending order
- [ ] Only includes games where the player's pick is alive
- [ ] `path_probability` is the product of individual game odds for each step
- [ ] Works with both brute-force and Monte Carlo engine results
- [ ] Returns empty `steps` list if player is eliminated or tournament is complete
- [ ] Known test: with current 2-player data, best path for trailing player should require specific remaining picks and have a calculable probability

## Branch

`feature/uc2-path-to-win`

### LINKED ISSUE #29: UC2-4: Plugin UI — "What needs to happen for me to win?" section in win_probability.py
## Summary

Wire the three new core helpers into `analyses/win_probability.py` as a new player-centric section. Users pick a player from a dropdown and see their critical games, best-case path, clinch status, and must-win flags — all in one place without navigating elsewhere.

## What to build

Extend `analyses/win_probability.py` with a new `_render_path_to_win(ctx, sr)` section rendered after the existing win probability table.

### UI layout

```
[ Player selector dropdown ]          ← defaults to my_player_name from config

--- If player is eliminated ---
  ⚠️  [Player] has been eliminated from contention.
      No combination of results can overcome the current gap.

--- If player is active ---
  🏆 Guaranteed Win                   ← only if clinch_scenarios() returns clinch_outcomes
      Root for: [Team] over [Team] in [Round]
      Root for: [Team] over [Team] in [Round]

  🎯 Your Best Path  (Win probability: X% | Path likelihood: Y%)
      1. Root for [Team] to beat [Team] — [Round]
      2. Root for [Team] to beat [Team] — [Round]
      ...

  ⚡ Games That Matter Most To You
      [Must-win badge] [Team A] vs [Team B] — [Round]
          Your win %: X% if [Team A] wins | 0% if [Team B] wins ← must-win
      [Team C] vs [Team D] — [Round]
          Your win %: X% if [Team C] wins | Y% if [Team D] wins
```

### Player selector behavior
- Dropdown if > 4 players, radio buttons if ≤ 4
- Default: `config.yaml → app.my_player_name`; falls back to rank-1 player
- Changing selection rerenders all three sections via Streamlit reactivity (no extra button needed)

### Must-win badge
- Show a red/orange "Must Win" label on any game where one outcome → 0% win probability for the selected player
- Sort must-win games first in the critical games list

## Key files

- `analyses/win_probability.py` — add `_render_path_to_win(ctx, sr)`, call from `render(ctx)`
- `core/scenarios.py` — consumes `player_critical_games()`, `clinch_scenarios()`, `best_path()`
- `config.yaml` — reads `app.my_player_name` (already added in UC1)

## Acceptance criteria

- [ ] Player selector defaults to `my_player_name` from config (falls back to rank-1)
- [ ] Eliminated players see the elimination message; path sections are suppressed
- [ ] "Guaranteed Win" section only appears when `clinch_outcomes` is not None
- [ ] Best path displays steps in round-ascending order with plain-English phrasing
- [ ] Path probability shown as a percentage alongside win probability
- [ ] Must-win games display first with visible badge
- [ ] Critical games re-ranked by selected player's swing (not global max swing)
- [ ] Switching player updates all sections without page reload
- [ ] Follows existing plugin rendering patterns (no new plugin file needed)

## Coordination note

When issue #22 (WI-5 counterfactual UI) is implemented, its player selector in `scenarios_whatif.py` should follow the same default-to-`my_player_name` pattern established here.

## Branch

`feature/uc2-path-to-win`

### PR DESCRIPTION
**Pull request by evan-firebrand**
## Requirements

Implement Use Case 2: "What needs to happen for me to win?" — the player-centric scenario analysis feature described in issues #25-#29. This includes:

- **UC2-1 (#26)**: `player_critical_games()` — re-rank critical games by personal swing, flag must-win games
- **UC2-2 (#27)**: `clinch_scenarios()` — deterministic clinch detection and minimum-picks-to-win calculation
- **UC2-3 (#28)**: `best_path()` — find the highest-probability winning path with odds-based probability
- **UC2-4 (#29)**: Plugin UI — player selector with personalized critical games, best path, and clinch status in `win_probability.py`

## Solution

Three files changed across core logic, UI, and tests:

- **`core/scenarios.py`** (+339 lines): Added `player_critical_games()`, `clinch_scenarios()`, `best_path()`, `_min_picks_to_lead()`, `_best_path_brute_force()`, `_best_path_greedy()`, and `GameProbability`/`CriticalGame` dataclasses. Dual-engine architecture: brute-force for <=15 remaining games, greedy swing-based selection for larger tournaments.
- **`analyses/win_probability.py`** (+127 lines): Added "What Needs to Happen for You to Win?" section with player selector (defaults to `my_player_name` from config), clinch detection, best-path instructions, must-win badges, and eliminated-player messaging.
- **`tests/test_scenarios.py`** (+170 lines): 18 new tests across `TestPlayerCriticalGames` (6), `TestClinchScenarios` (6), and `TestBestPath` (7) — all using deterministic 2-player fixtures.

## Issues & Revisions

- The original UC2 branch (`claude/priority-use-cases-ScKXa`) was based on main but had a follow-up fix commit addressing lazy import issues and a bug in `min_picks_needed` calculation where the scoring formula was not correctly accounting for round-based point values.
- The branch was rebased (cherry-picked) onto current main (`0f05ced`) on branch `claude/rebase-uc2-branch-ysnhL` to incorporate all merged work from PRs #23, #24, and #30.
- No merge conflicts were encountered during the rebase — the UC2 changes are additive and do not overlap with the counterfactual builder or bulk swap finder.

## Decisions

- **Dual-engine approach**: Brute-force enumeration for small tournaments (<=15 games) gives exact results; greedy swing-based selection handles larger tournaments. The threshold is configurable.
- **Odds transparency**: Every probability tracks its source (moneyline, spread, seed_historical, coin_flip) so users understand confidence levels.
- **Player-centric ranking**: Critical games are re-ranked per player rather than using global max_swing, which can highlight different games for different players.
- **Greedy vs exhaustive for best_path**: For large game counts, exhaustive enumeration is infeasible (2^N), so greedy selection by swing impact provides a reasonable approximation.

## Testing

- **Before**: 122 tests passing (22 scenario tests)
- **After**: 140 tests passing (40 scenario tests), +18 new tests
- All tests pass: `pytest` — 140 passed in ~5s
- Lint clean: `ruff check .` — all checks passed
- New test classes: `TestPlayerCriticalGames`, `TestClinchScenarios`, `TestBestPath` with deterministic fixtures

## Scope

All four sub-issues (#26-#29) are implemented as specified. No scope creep — the implementation stays within the defined UC2 boundary. The parent epic (#25) is fully addressed. No deferred work.

Closes #25, closes #26, closes #27, closes #28, closes #29

https://claude.ai/code/session_01UVWt5y1Bo1QxF57NbTZxji

### CODE REVIEW

File: core/scenarios.py [hunk 1]
```diff
@@ -625,6 +625,341 @@ def _build_critical_games_from_splits(
...
+    steps.sort(key=lambda s: s["round"])
+    return steps
+
+
+def _lowest_confidence(sources: list[str]) -> str:
+    """Return the lowest-confidence probability source from a list."""
+    order = ["coin_flip", "seed_historical", "spread", "moneyline"]
+    if not sources:
+        return "coin_flip"
+    return min(sources, key=lambda s: order.index(s) if s in order else len(order))
```
Comment (evan-firebrand) [inline core/scenarios.py]: **Edge case:** If a source string isn't in the `order` list, it gets `len(order)` as its index via the `else` clause, which ranks it as **highest** confidence. Unknown sources should probably be treated as **lowest** confidence instead:

```suggestion
    return min(sources, key=lambda s: order.index(s) if s in order else -1)
```

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by evan-firebrand — COMMENTED** [review]
## Review Summary

Good implementation of UC2. Clean separation of concerns — business logic in `core/scenarios.py`, presentation-only in `analyses/win_probability.py`. The fix commit (`3eb5f36`) properly extracts `_min_picks_to_lead()` as a reusable helper.

**CI:** ruff check passes, all 140 tests pass.

### Issues (3 inline comments)

1. **Bug (minor):** `clinch_scenarios()` line 738 returns `min_picks_needed: 0` when `alive_pending` is empty but `can_win=True` — should use `_min_picks_to_lead()` for consistency with the fix at line 759.
2. **Clarity:** Line 722 `hypo_results = results` — reference assignment is safe (due to `what_if()` immutability) but deserves a comment.
3. **Edge case:** `_lowest_confidence()` treats unknown source strings as highest confidence — should be lowest.

### Test gaps (1 inline comment)

- No direct tests for `_min_picks_to_lead()`, `_lowest_confidence()`, or the Monte Carlo greedy path. Nice-to-haves, not blockers.

### Positives

- Architecture follows the project's plugin pattern correctly
- 21 well-structured tests covering happy paths, edge cases, and eliminated players
- Defensive defaults (e.g., `sr.is_eliminated.get(player_name, True)`)
- Good PR description with all 6 required sections filled in substantively

---
_Generated by [Claude Code](https://claude.ai/code)_
